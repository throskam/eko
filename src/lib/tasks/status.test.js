jest.mock('../cio')
jest.mock('../config')
jest.mock('../git/status')

const cio = require('../cio')
const config = require('../config')
const gitStatus = require('../git/status')

const status = require('./status')

// Keep real implementation of message.
cio.message = jest.requireActual('../cio').message

beforeEach(() => {
  jest.resetAllMocks()
})

it('should display the project status', async () => {
  expect.assertions(1)

  const projects = [{
    directory: 'path/to/my-first-directory',
    repository: 'git@github.com:my-first-repository.git'
  }, {
    directory: 'path/to/my-second-directory',
    repository: 'git@github.com:my-second-repository.git'
  }, {
    directory: 'path/to/my-third-directory',
    repository: 'git@github.com:my-third-repository.git'
  }, {
    directory: 'path/to/my-fourth-directory',
    repository: 'git@github.com:my-fourth-repository.git'
  }]

  const messages = []

  config.projects.mockResolvedValue(projects)
  cio.wait.mockImplementation(p => p)

  gitStatus.mockResolvedValueOnce({
    directory: projects[0].directory,
    branch: 'short-name',
    state: '[ahead 1]',
    changes: 0
  })

  gitStatus.mockResolvedValueOnce({
    directory: projects[1].directory,
    branch: 'a-very-long-name-of-branch-that-is-awesome',
    state: '[behind 15]',
    changes: 1
  })

  gitStatus.mockResolvedValueOnce({
    directory: projects[2].directory,
    branch: 'some-other-branch',
    state: '',
    changes: 100
  })

  gitStatus.mockResolvedValueOnce({
    directory: projects[3].directory,
    error: new Error('Some error')
  })

  cio.log.mockImplementation(message => messages.push(message))

  await status()

  expect(messages).toMatchSnapshot()
})

it('should return when there is no projects', async () => {
  expect.assertions(1)

  config.projects.mockResolvedValue([])

  await status()

  expect(cio.wait).not.toHaveBeenCalled()
})
