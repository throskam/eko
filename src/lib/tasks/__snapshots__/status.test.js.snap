// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the project status 1`] = `
Array [
  "[36mâ–¶ [39mpath/to/my-first-directory [36m on [39mshort-name                                [36m [39m[ahead 1]  ",
  "[36mâ–¶ [39mpath/to/my-fourth-directory[36m stopped [39mSome error",
  "[36mâ–¶ [39mpath/to/my-second-directory[36m on [39ma-very-long-name-of-branch-that-is-awesome[36m [39m[behind 15][36m with [39m1[36m [39mchange",
  "[36mâ–¶ [39mpath/to/my-third-directory [36m on [39msome-other-branch                         [36m [39m           [36m with [39m100[36m [39mchanges",
]
`;
